contracts/dtoken/src/liquidation.rs

```diff
@@ -0,0 +1,126 @@
+use crate::*;
+
+#[near_bindgen]
+impl Contract {
+    pub fn liquidate(
+        &mut self,
+        borrower: AccountId,
+        borrowing_dtoken: AccountId,
+        liquidator: AccountId,
+        collateral_dtoken: AccountId,
+        liquidation_amount: WBalance,
+    ) -> PromiseOrValue<U128> {
+        assert_eq!(self.get_contract_address(), borrowing_dtoken);
+
+        controller::liquidation(
+            borrower.clone(),
+            borrowing_dtoken.clone(),
+            liquidator.clone(),
+            collateral_dtoken.clone(),
+            liquidation_amount,
+            self.get_controller_address(),
+            NO_DEPOSIT,
+            self.terra_gas(30),
+        )
+        .then(ext_self::liquidate_callback(
+            borrower,
+            borrowing_dtoken,
+            liquidator,
+            collateral_dtoken,
+            liquidation_amount,
+            env::current_account_id().clone(),
+            NO_DEPOSIT,
+            self.terra_gas(60),
+        ))
+        .into()
+    }
+
+    #[private]
+    pub fn liquidate_callback(
+        &mut self,
+        borrower: AccountId,
+        borrowing_dtoken: AccountId,
+        collateral_dtoken: AccountId,
+        liquidator: AccountId,
+        liquidation_amount: WBalance,
+    ) -> PromiseOrValue<U128> {
+        if !is_promise_success() {
+            Contract::custom_fail_log(
+                String::from("withdraw_fail"),
+                env::signer_account_id(),
+                Balance::from(liquidation_amount.0),
+                format!(
+                    "Liquidation failed, while tried to call controller ({:?})",
```

Comment:
@convowork1 we cannot use `:?` in production as it inteded for debug only
"Liquidation failed, while tried to call controller ({:?})" -> "failed to liquidate on {}", self.get_controller_address()